File: DeckOfCards-ObjC/DeckOfCardsTests/Classes/CardTests.m
//
//  CardTests.m
//  DeckOfCards
//
//  Created by Andrea Prearo on 11/20/13.
//  Copyright (c) 2013 Andrea Prearo. All rights reserved.
//

#import <XCTest/XCTest.h>
#import "Card.h"

@interface CardTests : XCTestCase

@end

@implementation CardTests

- (void)setUp
{
	[super setUp];
}

- (void)tearDown
{
	[super tearDown];
}

- (void)testInvalidCard
{
	XCTAssertThrowsSpecificNamed([[Card alloc] init], NSException, @"InvalidCardException",
															 @"A card needs to specify a suit and a rank");
}

- (void)testSuitCount
{
	XCTAssertTrue(numberOfSuits == 4, @"A card must define 4 possible suits");
}

- (void)testInvalidSuit
{
	XCTAssertThrowsSpecificNamed([[Card alloc] initWithSuit:-1 andRank:five], NSException,
															 @"InvalidCardException", @"Invalid suit");
	XCTAssertThrowsSpecificNamed([[Card alloc] initWithSuit:numberOfSuits andRank:five], NSException,
															 @"InvalidCardException", @"Invalid suit");
}

- (void)testRankCount
{
	XCTAssertTrue(numberOfRanks == 13, @"A card must define 13 possible ranks");
}

- (void)testInvalidRank
{
	XCTAssertThrowsSpecificNamed([[Card alloc] initWithSuit:clubs andRank:-1], NSException,
															 @"InvalidCardException", @"Invalid rank");
	XCTAssertThrowsSpecificNamed([[Card alloc] initWithSuit:clubs andRank:numberOfRanks], NSException,
															 @"InvalidCardException", @"Invalid rank");
}

@end

File: DeckOfCards-Swift/DeckOfCardsTests/Classes/CardTests.swift
//
//  CardTests.swift
//  DeckOfCards
//
//  Created by Andrea Prearo on 12/30/14.
//  Copyright (c) 2014 Andrea Prearo. All rights reserved.
//

import UIKit
import XCTest

class CardTests: XCTestCase {
    override func setUp() {
        super.setUp()
    }

    override func tearDown() {
        super.tearDown()
    }

    func testInvalidCard() {
        let card = Card(suit: Suit(rawValue: -100), rank: Rank(rawValue: -100))
        XCTAssertNil(card.suit, "A card should return Suit.Invalid for invalid suit values")
        XCTAssertNil(card.rank, "A card should return Rank.Invalid for invalid rank values")
    }

    func testSuitCount() {
        XCTAssertTrue(numberOfSuits == 4, "A card must define 4 possible suits")
    }

    func testRankCount() {
        XCTAssertTrue(numberOfRanks == 13, "A card must define 13 possible ranks")
    }

    func testEquality() {
        let card1 = Card(suit: .diamonds, rank: .queen)
        let card2 = Card(suit: .diamonds, rank: .queen)
        XCTAssertEqual(card1, card2)
        let card3 = Card(suit: .diamonds, rank: .ace)
        XCTAssertNotEqual(card1, card3)
    }
}
